### PR TITLE
[ResourceDetector.Host] Add host.id for non-containerized systems

### LINKED ISSUE #1630: Add host.id for non-containerized systems
# Issue with OpenTelemetry.ResourceDetectors.Host

Package `OpenTelemetry.ResourceDetectors.Host` version 0.1.0-alpha.2

**Is this a feature request or a bug?**

* [x] Feature Request
* [ ] Bug

**What is the expected behavior?**

`host.id` is a recommended attribute, and should be populated for non-containerized systems.

[The specification has a list of required sources](https://opentelemetry.io/docs/specs/semconv/resource/host/#collecting-hostid-from-non-containerized-systems) if host.id is collected outside containers.

**What is the actual behavior?**

`host.id` attribute is not populated for non-containerized systems.

## Additional Context

I'm available to implement this feature.

### PR DESCRIPTION
**Pull request by matt-hensley**
Fixes #1630 

## Changes

Adds support for setting `host.id` on Linux, MacOS, and Windows when running outside a container.

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.ResourceDetectors.Host/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+* Adds support for `host.id` resource attribute on non-containerized systems.
+`host.id` will be set per [semantic convention rules](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/host.md)
```
Comment (Kielek) [inline src/OpenTelemetry.ResourceDetectors.Host/CHANGELOG.md:6]:
```suggestion
`host.id` will be set per [semantic convention rules](https://github.com/open-telemetry/semantic-conventions/blob/v1.24.0/docs/resource/host.md)
```

Comment (Kielek) [reply src/OpenTelemetry.ResourceDetectors.Host/CHANGELOG.md:6]: Please fix compilation/linting issues.

**Comment by Kielek** [comment]
Please fix compilation/linting issues.

File: src/OpenTelemetry.ResourceDetectors.Host/HostDetector.cs [hunk 2]
```diff
@@ -33,4 +79,88 @@ public Resource Detect()
...
+    }
+
+#pragma warning disable CA1416
+    // stylecop wants this protected by System.OperatingSystem.IsWindows
+    // this type only exists in .NET 5+
+    private static string? GetMachineIdWindows()
+    {
+        try
+        {
+            return Registry.LocalMachine.OpenSubKey(@"SOFTWARE\Microsoft\Cryptography", false)?.GetValue("MachineGuid") as string ?? null;
```
Comment (lachmatt) [inline src/OpenTelemetry.ResourceDetectors.Host/HostDetector.cs:123]: `OpenSubKey` returns instance of `RegistryKey` that should be disposed.

File: src/OpenTelemetry.ResourceDetectors.Host/HostDetector.cs [hunk 3]
```diff
@@ -33,4 +79,88 @@ public Resource Detect()
...
+    }
+
+    private static string? GetMachineIdMacOs()
+    {
+        try
+        {
+            var startInfo = new ProcessStartInfo
+            {
+                FileName = "sh",
+                Arguments = "ioreg -rd1 -c IOPlatformExpertDevice | awk '/IOPlatformUUID/ { split($0, line, \"\\\"\"); printf(\"%s\\n\", line[4]); }'",
```
Comment (lachmatt) [inline src/OpenTelemetry.ResourceDetectors.Host/HostDetector.cs:96]: I'd prefer to have processing of results of `ioreg -rd1 -c IOPlatformExpertDevice` be done in-proc

File: src/OpenTelemetry.ResourceDetectors.Host/README.md [hunk 4]
```diff
@@ -37,6 +37,7 @@ The resource detectors will record the following metadata based on where
 your application is running:
 
 - **HostDetector**: `host.name`.
+- **HostDetector**: `host.id` when running on non-containerized systems.
```
Comment (Kielek) [inline src/OpenTelemetry.ResourceDetectors.Host/README.md:40]:
```suggestion
- **HostDetector**: `host.id` (when running on non-containerized systems), `host.name`.
```

**Review by Kielek — APPROVED** [review]

**Pull request merged**